Project 1 - DNA (Artem Zhdanov)

artemati opened the pull request.

Commit:
first commit

Commit:
Перенос категории

Перенес категорию в отдельный файл, как многие рекомендуют.

weitek commented on Project 1 - DNA/Project 1 - DNA/Cell+Mutator.m:
```diff
@@ -0,0 +1,41 @@
+    // Массив для хранения случайных индексов
+    NSMutableArray *rangeOfIndex = [NSMutableArray arrayWithCapacity:percentsForMutator];
+    // Так как массив у нас из 100 символов, то кол-во итераций равно кол-ву процентов
+    for (int i=0; i<percentsForMutator; i++) {
```
А что будет если кто-то сделает вызов этой процедуры с параметром больше 100?
Надо обязательно делать проверку на диапазон больше 0 и меньше или равно 100.
И действительно в этом проекте так замечательно совпало что кол-во элементов в массиве 100 и надо делать максимум сто операций замены, но как отлаживать алгоритм? Надо обязательно вводить или константу, или через #define - длина основного массива.
А здесь после проверки на диапазон допустимых значений для параметра процент изменений, делать расчет кол-ва изменений исходя из длины основного массива и процента необходимых изменений

artemati replied:
Согласен, обоснованное замечание!

weitek commented on Project 1 - DNA/Project 1 - DNA/Cell+Mutator.m:
```diff
@@ -0,0 +1,41 @@
+        while ([rangeOfIndex containsObject:indexNumber]) {
+            index = arc4random()%100;
+            indexNumber = [NSNumber numberWithInt:index];
+        }
```
Идея правильная, однако в реализации ошибка.
Если изначальные ДНК одинаковые и мы выполняем мутацию по одной на 100%, то результат различия не будет равен 100%. Хотя в алгоритме реализуется однозначное не соответствие заменяемого символа начальному символу.
Последовательный алгоритм - это не сумма отдельных строк и от перестановки местами результат изменяется. То есть если сейчас здесь переставить несколько строк выше, то все будет работать верно.

artemati replied:
Да, эту ошибку заметил уже после дедлайна. Так получилось из-за того, что сначала реализовывал замену символа, а потом замену по индексу, необходимо было делать наоборот. Ошибка налицо.

weitek replied:
Надо сделать исправления
После я ещё раз все проверю и выставлю оценки. Снижать оценку нет смысла так как идея реализована верно.

artemati replied:
Коммит с исправлениями сделал.

Commit:
FINAL

Учел все замечания.
Длину массива сделал через #define.
Алгоритм в Cell+Mutator.m подправил.
